### ISSUE TITLE
Strange SCF convergence with basis set projection with ECP

### ISSUE THREAD

**Issue by chenruduan** [psi4/psi4#1972]
Dear Psi4 developers,

We found strange SCF convergence behavior while the basis set projection with ECP is involved. This projection happens at scf_helper,
```
pCa = scf_wfn.basis_projection(Ca_occ, old_wfn.nalphapi(), old_wfn.basisset(), scf_wfn.basisset())
pCb = scf_wfn.basis_projection(Cb_occ, old_wfn.nbetapi(), old_wfn.basisset(), scf_wfn.basisset())
```
We think this function may not work as expected if `old_wfn.nalphapi() != scf_wfn.nalphapi()`. I posted it on the forum with my output files (http://forum.psicode.org/t/strange-scf-convergence-with-basis-set-projection-with-ecp/1885) and @hokru suggested I should open an issue here.

Best,
Chenru

**Comment by hokru** [comment]
Thanks.
The manual states basis set projection with ECPs is not tested, not sure if there are principle problem or if differently sized ECPs are causing this.
code section: https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/wavefunction.cc#L959

**Comment by susilehtola** [comment]
@chenruduan Please post the full output, not truncated versions. The input is not visible in the output you posted.

**Comment by susilehtola** [comment]
This issue is solvable, but I'm not sure how good the end result will be. The code needs to be changed so that it

1. determines the number of alpha and beta electrons in the new wave function
2. assembles the occupied orbitals in the old basis based on the number of electrons in the two calculations
2.1. if the number of electrons matches, everything is fine and the occupied orbitals are used as-is
2.2. if the new calculation has K fewer electrons, then the code must drop the K lowest-lying orbitals from the input
2.3. if the new calculation has K more electrons, then the guess fails, a warning message is printed out and the code switches to e.g. SAD
3. compute the projection with the given input orbitals

**Comment by susilehtola** [comment]
Scratch that, this is case 2.3: LANL2DZ uses a 1s2 2s2 2p6 frozen core on Co, while def2-SVP doesn't have a frozen core for Co. The projection would have to introduce the missing core orbitals to the guess. This is not an easy job. It would be interesting to know which is the "another software" that handles this...

**Comment by hokru** [comment]
I checkout how ORCA handles the projection and I see no obvious issues. The spin contamination after LANL2DZ->def2-SVP projection is tiny and SCF convergence normal.

They use a 'Fock-matrix projection' (FMatrix in the manual). Something Neese probably come up with during initial ORCA days and never published.

![image](https://user-images.githubusercontent.com/1648640/88648794-9da30f80-d0c7-11ea-8453-9e8cc3fe719b.png)

**Comment by susilehtola** [comment]
FWIW B3LYP/def2-SVP converges in 24 iterations from the default SAD guess, yielding S^2 = 0.7528. PBE/def2-SVP also converges in 25 iterations to S^2 = 0.7538. 

The weird convergence is a combination of the non-physical guess which lacks core electrons, combined with the unnecessary use of the second-order scheme which appears to have converged to an extremely high-lying saddle point: the log file on the forum has a B3LYP/def2-SVP energy of -1803.88405461301363 after 77 SOKS iterations, while I get -1820.53977771750078, which is SEVENTEEN HARTREE lower....

**Comment by susilehtola** [comment]
@hokru Fock matrix projection is a well-known method; you just project the Fock matrix into the new basis and diagonalize it there.

However, it will also fail in this case since the core electrons  j u s t  a r e  n o t  t h e r e....

**Comment by susilehtola** [comment]
IMHO the best solution would just be to turn off the projection if the number of electrons doesn't match. Projecting between different ECPs is not going to yield good results. The calculation works just fine from scratch.

**Comment by hokru** [comment]
I tend to agree that turning off the projection in that case is the most sensible.

The fock matrix projection in ORCA seems to 'work' in the sense that it convergences to the same result as the default guess (PModel) in some fewer iterations. The guess energy is quite far off, though. General robustness is thus questionable.

**Comment by susilehtola** [comment]
I don't think this has anything to do with the projection; but rather the use of the second-order convergence acceleration scheme. Orca's `FMatrix` guess also misses the core electrons (since they don't exist in the orbitals), but the first diagonalization should reintroduce them. Of course, a question is, how is the projection implemented in Orca? You can either project the full Fock matrix, in which case the null space will be zero, or you could just project the XC part in which case the null space will essentially be the core Hamiltonian. The latter scheme would restore the core orbitals, but they wouldn't be screened at all; you might also introduce spurious valence states which will ruin the performance of the guess...

Probably the best scheme would be to do the mixed-basis version in which you compute the Fock matrix in the new basis by the orbitals in the old basis without any projection operators. However, this method requires the capability to compute mixed-basis integrals which is not available in Psi4.

**Comment by chenruduan** [comment]
Thanks for the replies! Just FYI that the "other software" that I tested, which seemingly worked (at least converged to a reasonable energy with the similar spin-square value) are ORCA and Terachem, as you may wonder. And yes, calculations from scratch for both basis sets work can converge to the same results. The issue only happens when I tried to do the project as I want to ensure the convergence to the same electronic state.

**Comment by susilehtola** [comment]
Well, the fix I proposed above in https://github.com/psi4/psi4/issues/1972#issuecomment-664898846 would allow going from a converged def2-SVP to LANL and thus ensure the electronic states are the same. Going the opposite way doesn't seem as straightforward; I think it would require the mixed-basis approach.

**Comment by PeterKraus** [comment]
I'm surprised the parser actually allows for this at all. The "clever" cast-up basis selection was implemented in #1484, but if the default is over-ridden, it's hands off.

Whoever fixes this should probably check whether the values in [this test](https://github.com/psi4/psi4/blob/master/tests/basis-ecp/input.dat) are sensible.
